musicgen/musicgen.py

```diff
@@ -0,0 +1,359 @@
+    probs = mx.softmax(logits * (1 / temperature), axis=axis)
+
+    # sort probs in ascending order
+    sorted_indices = mx.argsort(probs, axis=axis)
+    sorted_probs = mx.take_along_axis(probs, sorted_indices, axis=axis)
+    prob_threshold = mx.take(sorted_probs, mx.array(-top_k), axis=axis)
+
+    # select the top K tokens in probability
+    top_probs = mx.where(
+        sorted_probs > prob_threshold,
+        sorted_probs,
+        0,
+    )
+
+    sorted_token = mx.random.categorical(mx.log(top_probs), axis=axis)
+    token = mx.take_along_axis(
+        sorted_indices, sorted_token[:, :, mx.newaxis], axis=axis
+    )
+
+    return token
+
+
+def create_sin_embedding(positions: mx.array, dim: int, max_period: float = 10000):
+    assert dim % 2 == 0
+    half_dim = dim // 2
+    adim = mx.arange(half_dim).reshape(1, 1, -1)
+    phase = positions / (max_period ** (adim / (half_dim - 1)))
+    return mx.concatenate([mx.cos(phase), mx.sin(phase)], axis=-1)
+
+
+class MusicGen(nn.Module):
+    def __init__(self, config):
+        self.num_codebooks = config.decoder.num_codebooks
+        self.codebook_size = config.audio_encoder.codebook_size
+        self.bos_token_id = config.decoder.bos_token_id
+        self.hidden_size = config.decoder.hidden_size
+        self.num_attention_heads = config.decoder.num_attention_heads
+        self.sampling_rate = config.audio_encoder.sampling_rate
+
+        self.text_conditioner = TextConditioner(
+            config.text_encoder._name_or_path,
+            config.text_encoder.d_model,
+            self.hidden_size,
+        )
+        self.emb = [
+            nn.Embedding(self.codebook_size + 1, self.hidden_size)
+            for _ in range(self.num_codebooks)
+        ]
+        self.layers = [
+            TransformerBlock(config) for _ in range(config.decoder.num_hidden_layers)
+        ]
+        self.out_norm = nn.LayerNorm(self.hidden_size, eps=1e-5)
+        self.linears = [
+            nn.Linear(self.hidden_size, self.codebook_size, bias=False)
+            for _ in range(self.num_codebooks)
+        ]
+        encodec_name = config.audio_encoder._name_or_path.split("/")[-1]
+        encodec_name = encodec_name.replace("_", "-")
+        self._audio_decoder, _ = EncodecModel.from_pretrained(
+            f"mlx-community/{encodec_name}-float32"
+        )
+
+    def __call__(
+        self,
+        audio_tokens: mx.array,
+        conditioning: mx.array,
+        cache: list[KVCache] = None,
+    ):
+
+        if cache is None:
+            cache = [None] * len(self.layers)
+
+        x = sum([self.emb[k](audio_tokens[..., k]) for k in range(self.num_codebooks)])
+
+        offset = cache[0].offset if cache[0] is not None else 0
+        offset = mx.full((audio_tokens.shape[0], 1, 1), offset)
+        pos_emb = create_sin_embedding(offset, self.hidden_size)
+        x += pos_emb
```

Comment:
II'm not sure if the models are ever in lower precision. If so a cast here will be important: `x += pos_emb.astype(x.dtype)`